Team Route Patches

rayan-hamid opened the pull request:
# Description

This pull request secures all relevant routes by adding proper authentication middleware where it was previously missing. The changes were made collaboratively by @rayan-hamid, @misantiago99, and @cachan3543 to ensure that all access points are now properly protected based on project discussions.

## Type of change

- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.org/developer/pr/) ideas
- [x] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.org/developer/cla/) and each author is listed in the Description section.

## Limitations

Further testing may be helpful to confirm that all routes behave securely in various edge cases and real-world scenarios.

rayan-hamid committed:
Added authentication middleware to conversions and conversionArray routes

misantiago99 committed:
Patch access control for route files  groups, login, logs, maps, meters, obvious, and preferences

misantiago99 committed:
Applying fixes that Steve suggested. Reverted docker.compose.yml and package-lock.js by checking out from development branch and overriding anything that I could have done to those files.

rayan-hamid committed:
Merge remote-tracking branch 'origin/Miguel---Testing' into Rayan-patches

rayan-hamid committed:
Merge remote-tracking branch 'upstream/development' into Rayan-patches

rayan-hamid committed:
Merge remote-tracking branch 'origin/Calvin-patches' into Rayan-patches

rayan-hamid committed:
Pulled in updates from Miguel and Calvin: maps.js, login.js, package.json, readings.js, and units.js

huss commented on src/server/routes/conversionArray.js:
```diff
@@ -6,13 +6,14 @@ const express = require('express');
  * Route for redoing Cik and/or refreshing reading views.
  */
-router.post('/refresh', async (req, res) => {
+router.post('/refresh', adminAuthMiddleware('refresh system data'), async (req, res) => {
```
Could the description become 'refresh conversions array' so it has the actual route in it as other descriptions.

huss commented on src/server/routes/unitReadings.js:
```diff
@@ -377,7 +390,10 @@ function validateThreeDQueryParams(queryParams) {
+  /**
+   * Route for fetching line readings by meter IDs
+   */
+	router.get('/line/meters/:meter_ids', authMiddleware('view readings'), async (req, res) => {
```
All the descriptions in this file have the same wording. It should be unique to each route. I thought I saw an update on a branch that addressed this. Maybe it needs to be merged in? Let me know if I'm missing something.

huss commented on src/server/routes/units.js:
```diff
@@ -1,258 +1,146 @@
+ const { success, failure } = require('./response');
+ 
+ const router = express.Router();
+ 
```
The formatUnitForResponse function was removed. I realize it is only used once but it does abstract the idea. Could it please be added back.

huss commented on src/server/routes/units.js:
```diff
@@ -1,258 +1,146 @@
-	}
-});
-
+ router.get(
```
The formatting was changed so this is now multiline. It does not match other files and causes more indenting. Could it please be done the other way.

Also, the indenting is now with spaces instead of tabs. VSC an change this throughout the file. Ask if you need help. Note that here there should be no indenting here,

huss commented on src/server/routes/units.js:
```diff
@@ -1,258 +1,146 @@
-
+ router.post(
+   '/edit',
+   authMiddleware('make changes to a unit'),
```
Please make the description 'edit units'.

huss commented on src/server/routes/units.js:
```diff
@@ -1,258 +1,146 @@
-
+ router.get(
+   '/',
+   authMiddleware('manage units'),
```
Please make description 'get units'.

huss commented on src/server/routes/units.js:
```diff
@@ -1,258 +1,146 @@
-	}
-});
-
+ router.post(
```
There are a lot of changes in the remainder of this file including removal of the request parameter validation, some comments (some changes might be okay but others are not) and response messages. I'm unclear why they were made. Could you either explain of put them back. I'll need to review this file again once this is settled.

huss replied:
I resolved this by removing most changes.

huss commented on src/server/routes/units.js:
```diff
@@ -1,258 +1,146 @@
-
+ router.post(
+   '/addUnit',
+   authMiddleware('add a new unit'),
```
The description should be 'add units'.

huss commented on src/server/routes/units.js:
```diff
@@ -1,258 +1,146 @@
-module.exports = router;
+ router.post(
+   '/delete',
+   authMiddleware('removes a certain unit'),
```
'delete units'

huss commented on src/server/routes/users.js:
```diff
@@ -135,7 +137,7 @@ router.post('/create', adminAuthMiddleware('create a user.'), async (req, res) =
- * Route for updating user role
+ * Route for updating an existing user.
  */
 router.post('/edit', adminAuthMiddleware('update a user role'), async (req, res) => {
```
I realize you did not do this but could you change description to 'edit a user'.

huss reviewed (changes requested):
Thanks to @rayan-hamid, @misantiago99, and @cachan3543 for this contribution. I've made several comments to address. Please let me know if you have questions or need help.

rayan-hamid committed:
Resolved unitReadings.js conflict by using Calvin's updated version

rayan-hamid committed:
Resolved all reviewer feedback: route descriptions, validation schemas, formatting

rayan-hamid committed:
Updated route descriptions and authorization roles in users.js and conversionArray.js per PR feedback

huss commented on src/server/routes/conversionArray.js:
```diff
@@ -6,13 +6,14 @@ const express = require('express');
  * Route for redoing Cik and/or refreshing reading views.
  */
-router.post('/refresh', async (req, res) => {
+router.post('/refresh conversions array', adminAuthMiddleware('refresh system data'), async (req, res) => {
```
I think the name of the route and not the description was updated. The route name should be put back and the adminAuthMiddleware parameter changed to the new value.

huss replied:
I fixed this.

huss commented on src/server/routes/units.js:
```diff
@@ -1,258 +1,186 @@
-const { removeAdditionalConversionsAndUnits } = require('../services/graph/handleSuffixUnits');
-const validate = require('jsonschema').validate;
-const { success, failure } = require('./response');
+ const express = require('express');
```
The formatting is off with spaces at the start of lines and the use of spaces instead of tabs for indenting. VSC can format a document to fix this.

huss replied:
I fixed this.

huss commented on src/server/routes/meters.js:
```diff
@@ -103,7 +102,9 @@ function formatMeterForResponse(meter, hasFullAccess) {
  */
-router.get('/', async (req, res) => {
+
+//Adding optionalAuthMiddlewear directly to route
```
I think the "Adding ...." comment can be removed from each route in this file. OED does not need a comment to note changes.

huss replied:
I fixed this.

huss commented on src/server/routes/units.js:
```diff
@@ -1,258 +1,186 @@
+// routes/units.js
```
I think this comment needs to be removed. It certainly should not be a the top of the file before the license.

huss replied:
I resolved this.

huss commented on src/server/routes/units.js:
```diff
@@ -1,258 +1,186 @@
+	 try {
+	   const unit = await Unit.getById(req.body.id, conn);
+	   if (unit.suffix !== req.body.suffix) {
+		 // Remove old conversions if suffix has changed
```
I'm unclear on why this comment was changed, esp. since it does not relate to the work done.

huss replied:
I resolved this.

huss reviewed (changes requested):
@rayan-hamid, @misantiago99, and @cachan3543 Thanks for the updated code and sorry it took a while to review. There are still some comments to resolve including a few from before, new ones relating to previous ones and a couple of new things I just noticed. Could you either address them or let me know if you want help.

huss commented:
@rayan-hamid, @misantiago99, and @cachan3543 It has been over a month since my last comments. OED appreciates your work so far. OED would like to move this work forward. If you are working on this or need help then please leave a comment. If nothing is posted by 1 Aug. then OED will move forward to complete this work.

huss committed:
This changes several items in routes:
- Standardize names/imports to adminAuthMiddleware, optionalAuthMiddleware
- Fixed some optionalAuthMiddleware that should not receive a parameter
- Removed usage of authMiddleware as exported function so all use one above
- Fixed units edit, add & delete routes to require admin
- Put in some TODOs
- Minor fixes/formatting
- The above addressed some PR comments

huss committed:
fix undesired changes to unit.js route

huss committed:
address two PR comments

huss committed:
Merge remote-tracking branch 'origin/development' into pr/rayan-hamid/1489

huss committed:
minor edit from reading

Mostly comments.

huss commented:
I have made a number of commits. These address PR comments, make desired changes & generally get PR ready for merging.

huss reviewed (approved):
Per my comment, I have updated the PR so it is ready to merge.
